### PR TITLE
2nd PR of Python/GT4Py implementation (ShallowWaterState).

### PR DESCRIPTION
**Pull request by robgpita**
## Proposed changes

This is the second of ~5 PRs that will introduce a Python/GT4Py implementation of the Shallow Water code.

This PR introduces the Python Class implementation of ShallowWaterState and associated Pytest file.
 
Tested on laptop with 1,2,4, & 9 processes:
`mpiexec -n 9 --oversubscribe python -m pytest --with-mpi -s --disable-pytest-warnings`

And for non-mpi tests:
`mpiexec -n 1 python -m pytest`

This PR also includes some changes to [.github/workflows/macos_intel.yml](https://github.com/NOAA-GSL/SENA-shallow-water/compare/feature/python-state?expand=1#diff-615c12e4300e83697176d001cde7564689560b65caa79aeefa15e21455122b41) in an attempt to resolve the  C++ compiler used by the GHA runner when buildlng Serialbox and update `actions/cache@v2` to `actions/cache@v3`

## Types of changes

What types of changes does your code introduce?

- [ ] Bugfix (Change which fixes an issue)
- [x] New feature (Change which adds functionality)
- [x] Enhancement to existing functionality (non-breaking change which improves existing code or documentation)
- [ ] Breaking change (fix or feature that changes results and requires updates to test baselines)
- [ ] Documentation Update

## Checklist

_Replace the space with an `x` for ALL the boxes that apply._

_You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the [CONTRIBUTING](https://github.com/NOAA-GSL/SENA-shallow-water/blob/develop/CONTRIBUTING.md) document
- [x] I have ensured all my changes contribute to a common purpose
- [x] I have verified my change does not add debug code
- [x] I have verified my change does not add code that is commented out
- [x] I have verified unit tests pass locally with my changes
- [x] I have verified my changes adhere to style guidelines
- [x] I have added tests that prove my fix is effective or that my feature works (if applicable)
- [ ] I have added necessary documentation (if appropriate)

## Reviewers' checklist

The following is a checklist for reviewers to consider when reviewing this pull request.

#### Implementation

- [ ] Does this code do what it is supposed to do?
- [ ] Can the code be simplified?
- [ ] Does the code add unwanted compile-time or run-time dependencies?
- [ ] Is the code at the correct abstraction level?
- [ ] Is the code modular enough?
- [ ] Does similar functionality already exist in the codebase? If so, why wasn't it used?
- [ ] Can you think of a different solution that is substantially better in terms of code maintenance, readability, or performance?

#### Logic Errors and Bugs

- [ ] Can you think of a use case in which the code does not behave properly?
- [ ] Can you think of any inputs or external events that could break the code?

#### Error handling and logging

- [ ] Is error handling done correctly?
- [ ] Should any logging or debuggging be added or removed?
- [ ] Are error messages user friendly?

#### Usability

- [ ] Is the API intuitive to use?
- [ ] Is the API well documented?

#### Testing and testability

- [ ] Do the tests pass?
- [ ] Is the code testable?
- [ ] Do the tests reasonably cover the code change?
- [ ] Are there additional edge cases or inputs that should be tested?

#### Performance

- [ ] Do you think the code change will negatively impact performance?
- [ ] Do you see any potential to improve the performance of the code?

#### Dependencies

- [ ] Does this change require updates to documenation, configuration?  If so, what that done?
- [ ] Does this change affect backward compatibility?

#### Readability

- [ ] Is the code easy to understand?
- [ ] Can the readability be improved with smaller routines?
- [ ] Can the readability be improved with different subroutine/variable names?
- [ ] Is the data flow understandable?
- [ ] Could comments in the code be improved to add clarity?
- [ ] Is there any commented out code?
- [ ] Does the code conform to the style guide (Fortran style compliance cannot be checked automatically)

### CODE REVIEW

File: test/python/test_shallow_water_state.py [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+    state.write("../test_output/test_shallow_water_writer.nc")
+
+    state_read = ShallowWaterState(input_io_geometry, input_gt4py_config)
+    state_read.read("../test_output/test_shallow_water_writer.nc")
+
+    assert_almost_equal(state.clock, state_read.clock, decimal=9)
+
+    assert_array_equal(state_read.u.data[xps-xms:xpe-xms+1, yps-yms:ype-yms+1], u)
+    assert_array_equal(state_read.v.data[xps-xms:xpe-xms+1, yps-yms:ype-yms+1], v)
+    assert_array_equal(state_read.h.data[xps-xms:xpe-xms+1, yps-yms:ype-yms+1], h)
```
Comment (christopherwharrop-noaa) [inline test/python/test_shallow_water_state.py]: Need a newline here

**Review by christopherwharrop-noaa — CHANGES_REQUESTED** [review]
One tiny syntax thing.

File: test/python/test_shallow_water_state.py (hunk 1 above)
Comment (robgpita) [reply test/python/test_shallow_water_state.py]: @christopherwharrop-noaa I just caught that as well and added the newline.

**Review by christopherwharrop-noaa — CHANGES_REQUESTED** [review]
The new state test fails because it uses `../test_output/` to read/write files for the NetCDF I/O tests.  Once you create an empty test_output directory (you'll have to do `touch .gitignore` in there to make git save an empty dir), it should work.  After you add that, we should be good to go.

**Comment by robgpita** [comment]
Right. Thanks for catching this. Those tests were passing locally, but of course because I have `test/test_output`. Adding the `.gitignore` file into the empty directory does the trick.

**Comment by christopherwharrop-noaa** [comment]
I am cloning a fresh checkout of the PR branch each time and manually running the pytests.  Once we have the pytests all working in the CI, we won't need to do that.

**Review by christopherwharrop-noaa — APPROVED** [review]
